packaging/metadata.py

```diff
@@ -0,0 +1,416 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+import dataclasses
+import math
+import sys
+import textwrap
+from email import message_from_bytes
+from email.contentmanager import raw_data_manager
+from email.headerregistry import Address, AddressHeader
+from email.message import EmailMessage
+from email.policy import EmailPolicy, Policy
+from functools import lru_cache, reduce
+from itertools import chain
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ClassVar,
+    Collection,
+    Dict,
+    Iterable,
+    Iterator,
+    Mapping,
+    NamedTuple,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from .requirements import InvalidRequirement, Requirement
+from .specifiers import SpecifierSet
+from .version import Version
+
+T = TypeVar("T", bound="CoreMetadata")
+A = TypeVar("A")
+B = TypeVar("B")
+
+if sys.version_info[:2] >= (3, 8) and TYPE_CHECKING:  # pragma: no cover
+    from typing import Literal
+
+    NormalizedDynamicFields = Literal[
+        "platform",
+        "summary",
+        "description",
+        "keywords",
+        "home-page",
+        "author",
+        "author-email",
+        "license",
+        "supported-platform",
+        "download-url",
+        "classifier",
+        "maintainer",
+        "maintainer-email",
+        "requires-dist",
+        "requires-python",
+        "requires-external",
+        "project-url",
+        "provides-extra",
+        "provides-dist",
+        "obsoletes-dist",
+        "description-content-type",
+    ]
+else:
+    NormalizedDynamicFields = str
+
+
+def _normalize_field_name_for_dynamic(field: str) -> NormalizedDynamicFields:
+    """Normalize a metadata field name that is acceptable in `dynamic`.
+
+    The field name will be normalized to lower-case. JSON field names are
+    also acceptable and will be translated accordingly.
+
+    """
+    return cast(NormalizedDynamicFields, field.lower().replace("_", "-"))
+
+
+# Bypass frozen dataclass for __post_init__, this approach is documented in:
+# https://docs.python.org/3/library/dataclasses.html#frozen-instances
+_setattr = object.__setattr__
+
+
+class EmailAddress(NamedTuple):
```

Comment:
We're not supporting an old tuple-based API, so there's no need for this to be a named tuple and have to support two different APIs for as long as this code exists.

But also, why the custom class? Convenience?

Reply:
Thank you very much for the comment, please allow me to reply backwards:

> why the custom class? Convenience?

There are 2 interconnected aspects on having a custom class. As you mentioned it is convenient, specially because of the `__str__` method that allows a correct serialisation of the given email address (with or without `display_name`), and it is used in the `to_*` methods. The second aspect, and the main reason why I did not use the stdlib-provided class `email.headerregistry.Address` directly, is that `frozenset` requires hashable objects. So having a `NamedTuple` with a custom `__str__` method replicating what `email.headerregistry.Address` does made sense to me when I was writing this code.


> We're not supporting an old tuple-based API, so there's no need for this to be a named tuple and have to support two different APIs for as long as this code exists.

I am sorry, I don't know if I understand what you mean here.
Do you mean maintaining the `EmailAddress` interface as part of the API? If that is the case, could I remove it from the dataclass signature and make it private while still relying on it as a implementation detail. Or would it be better to just use a regular tuple and create a private helper method just meant to be used for the serialisation? Do you have any other suggestion?

Reply:
I would do the latter; keep it a tuple and have a private function/method for serialization.

Reply:
Thank you very much, I have updated the implementation following this suggestion.